lock-api/src/main/java/com/palantir/lock/client/LeaderElectionReportingTimelockService.java

```diff
@@ -122,6 +140,129 @@ public ConjureStartTransactionsResponse startTransactions(ConjureStartTransactio
         return election && (currentLeader != null);
     }
 
+    /**
+     * Estimating the duration of leader election:
+     *
+     * If a request is sent out at time T_1, and a response arrives at time T_2 with the id of leader A, we can
+     * deduce three things:
+     *   1. A could not have lost leadership before T_1 if it was the leader before it (upper bound)
+     *   2. A must have gained leadership before T_2 (lower bound)
+     *   3. A was the leader at least in some point in the interval [T_1, T_2]
+     * Although from just one request, the lower bound is greater than the upper bound, after multiple responses that
+     * is generally not going to be the case anymore.
+     *
+     * Ordering leaderships:
+     * Let L_A and U_A be the lower and upper bound, respectively, of leader A, and let L_B and U_B be the respective
+     * bounds of leader B. We distinguish 3 cases:
+     *   1. (L_A < U_A && L_B < U_B): for both leaders there is an interval of guaranteed leadership, which by
+     *   definitions cannot overlap. It is therefore trivial to order leadership of A and B.
+     *   2. (w.l.o.g. L_A < U_A && L_B >= U_B):
+     *      a) if U_B > U_A, A was the leader before B
+     *      b) if L_B < L_A, B was the leader before A
+     *      c) otherwise, we cannot determine the ordering
+     *   3. (L_A >= U_A && L_B >= U_B):
+     *      a) if U_B > L_A, A was the leader before B
+     *      b) if U_A > L_B, B was the leader before A
+     *      c) otherwise, we cannot determine the ordering
```

Comment:
I _think_ this is right, but was also pretty hard to reason through - I had to write it out as above to verify that this works.

We're definitely going to want some solid ETE tests here to make sure that this is right. Fortunately, the cost of failure here is just dodgy metrics, so an ETE test in a separate PR is fine, but we'll still want to verify this logic.

Reply:
That's not how mathematical proofs work

Reply:
I know, but it would be nice to verify that your code actually works as intended here (unless you want to write a proof that your code is the same here too?)